server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java

```diff
@@ -486,4 +539,323 @@ public void testShardLimit() {
         assertThat(e, hasToString(containsString(expectedMessage)));
     }
 
+    public void testParseMappingsAppliesDataFromTemplateAndRequest() throws Exception {
+        IndexTemplateMetaData templateMetaData = addMatchingTemplate(templateBuilder -> {
+            templateBuilder.putAlias(AliasMetaData.builder("alias1"));
+            templateBuilder.putMapping("_doc", createMapping("mapping_from_template", "text"));
+        });
+        request.mappings(createMapping("mapping_from_request", "text").string());
+
+        Map<String, Object> parsedMappings = MetaDataCreateIndexService.parseMappings(request.mappings(), List.of(templateMetaData),
+            NamedXContentRegistry.EMPTY);
+
+        assertThat(parsedMappings, hasKey("_doc"));
+        Map<String, Object> doc = (Map<String, Object>) parsedMappings.get("_doc");
+        assertThat(doc, hasKey("properties"));
+        Map<String, Object> mappingsProperties = (Map<String, Object>) doc.get("properties");
+        assertThat(mappingsProperties, hasKey("mapping_from_request"));
+        assertThat(mappingsProperties, hasKey("mapping_from_template"));
+    }
+
+    public void testAggregateSettingsAppliesSettingsFromTemplatesAndRequest() {
+        IndexTemplateMetaData templateMetaData = addMatchingTemplate(builder -> {
+            builder.settings(Settings.builder().put("template_setting", "value1"));
+        });
+        ImmutableOpenMap.Builder<String, IndexTemplateMetaData> templatesBuilder = ImmutableOpenMap.builder();
+        templatesBuilder.put("template_1", templateMetaData);
+        MetaData metaData = new MetaData.Builder().templates(templatesBuilder.build()).build();
+        ClusterState clusterState = ClusterState.builder(org.elasticsearch.cluster.ClusterName.CLUSTER_NAME_SETTING
+            .getDefault(Settings.EMPTY))
+            .metaData(metaData)
+            .build();
+        request.settings(Settings.builder().put("request_setting", "value2").build());
+
+        Settings aggregatedIndexSettings = aggregateIndexSettings(clusterState, request, List.of(templateMetaData), Map.of(),
+            null, Settings.EMPTY, IndexScopedSettings.DEFAULT_SCOPED_SETTINGS);
+
+        assertThat(aggregatedIndexSettings.get("template_setting"), equalTo("value1"));
+        assertThat(aggregatedIndexSettings.get("request_setting"), equalTo("value2"));
+    }
+
+    private IndexTemplateMetaData addMatchingTemplate(Consumer<IndexTemplateMetaData.Builder> configurator) {
```

Comment:
Can you put the test helpers below all the `testXYZ` methods? It's harder to read when they're scattered throughout the file